streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopicManager.java

```diff
@@ -63,83 +93,172 @@ public InternalTopicManager(final StreamsKafkaClient streamsKafkaClient,
+                        if (createTopicNames.contains(it.next().name())) {
+                            it.remove();
+                        }
                     }
-                    streamsKafkaClient.createTopics(topicsToBeCreated, replicationFactor, windowChangeLogAdditionalRetention, metadata);
+
+                    continue;
                 }
+
                 return;
-            } catch (StreamsException ex) {
-                log.warn("Could not create internal topics: {} Retry #{}", ex.getMessage(), i);
-            }
-            // backoff
-            time.sleep(100L);
+            } while (remainingRetries-- > 0);
+
+            final String timeoutAndRetryError = "Could not create topics. " +
+                "This can happen if the Kafka cluster is temporary not available. " +
+                "You can increase admin client config `retries` to be resilient against this error.";
+            log.error(timeoutAndRetryError);
+            throw new StreamsException(timeoutAndRetryError);
         }
-        throw new StreamsException("Could not create internal topics.");
     }
 
     /**
      * Get the number of partitions for the given topics
      */
     public Map<String, Integer> getNumPartitions(final Set<String> topics) {
-        for (int i = 0; i < MAX_TOPIC_READY_TRY; i++) {
-            try {
-                final MetadataResponse metadata = streamsKafkaClient.fetchMetadata();
-                final Map<String, Integer> existingTopicPartitions = fetchExistingPartitionCountByTopic(metadata);
-                existingTopicPartitions.keySet().retainAll(topics);
-
-                return existingTopicPartitions;
-            } catch (StreamsException ex) {
-                log.warn("Could not get number of partitions: {} Retry #{}", ex.getMessage(), i);
+        return getNumPartitions(topics, false);
+    }
+
+    private Map<String, Integer> getNumPartitions(final Set<String> topics,
+                                                  final boolean bestEffort) {
+        int remainingRetries = retries;
+        boolean retry;
+        do {
+            retry = false;
+
+            final DescribeTopicsResult describeTopicsResult = adminClient.describeTopics(topics);
+            final Map<String, KafkaFuture<TopicDescription>> futures = describeTopicsResult.values();
+
+            final Map<String, Integer> existingNumberOfPartitionsPerTopic = new HashMap<>();
+            for (final Map.Entry<String, KafkaFuture<TopicDescription>> topicFuture : futures.entrySet()) {
+                try {
+                    final TopicDescription topicDescription = topicFuture.getValue().get();
+                    existingNumberOfPartitionsPerTopic.put(
+                        topicFuture.getKey(),
+                        topicDescription.partitions().size());
+                } catch (final InterruptedException fatalException) {
+                    Thread.currentThread().interrupt();
```

Comment:
Ditto above, if we do not expect this to ever happen, we do not need to `Thread.currentThread().interrupt();` and instead just log error and throw as IllegalState with the error message indicating this should never happen.

Reply:
It doesn't hurt to set the interrupted flag -- and it's correct behavior. So I would prefer it to keep it. Will update the error msg.